avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -71,4 +78,35 @@ protected Object deserialize(byte[] payload) throws SerializationException {
       throw new SerializationException("Error deserializing Avro message for id " + id, e);
     }
   }
+
+  private DatumReader getDatumReader(Schema writerSchema) {
+    if (useSpecificAvroReader) {
+      return new SpecificDatumReader(writerSchema, getReaderSchema(writerSchema));
+    } else {
+      return new GenericDatumReader(writerSchema);
+    }
+  }
+
+  private Schema getReaderSchema(Schema writerSchema) {
+    Schema readerSchema = targetSchemaCache.get(writerSchema);
+    if (readerSchema == null) {
+      Class targetClass = SpecificData.get().getClass(writerSchema);
+      if (targetClass != null) {
+        try {
+          readerSchema = ((SpecificRecord) targetClass.newInstance()).getSchema();
+        } catch (InstantiationException e) {
+          throw new SerializationException(writerSchema.getFullName() + " specified by the " +
+            "writers schema is not a SpecificRecord which is required to find the readers schema.");
+        } catch (IllegalAccessException e) {
+          throw new SerializationException(writerSchema.getFullName() + " specified by the " +
+            "writers schema is not a SpecificRecord which is required to find the readers schema.");
```

Comment:
Could we distinguish the message string caused by InstantiationException and IllegalAccessException?